GBA/bitmap.zig

```diff
@@ -1,106 +1,114 @@
-                if (dd >= 0) {
-                    dd -= 2 * dy;
-                    destinationIndex += xstep;
-                }
+        }
 
-                dd += 2 * dx;
-                destinationIndex += ystep;
+        // TODO: Test using full word memcpy.
+        inline fn lineHorizontal(x1: u8, x2: u8, y: u8, color: Color) void {
+            if (@sizeOf(Color) == 2) {
+                for (screen()[y][x1 .. x2 + 1]) |*pixel| pixel.* = color;
+            } else {
+                const first = x1 >> 1;
+                const last = x2 >> 1;
+                const full: Color = .{ .lo = color, .hi = color };
+                if (x1 & 1 == 0)
+                    screen()[y][first] = full
+                else
+                    screen()[y][first].hi = color;
+                for (screen[y][first + 1 .. last]) |*x| x.* = full;
+                if (x2 & 1 == 0)
+                    screen()[y][last].lo = color
+                else
+                    screen()[y][last] = full;
             }
         }
-    }
-
-    pub fn rect(left: i32, top: i32, right: i32, bottom: i32, color: u16, destinationBase: [*]volatile u16, rawPitch: i32) void {
-        var ix: i32 = 0;
-        var iy: i32 = 0;
-        const width: i32 = right - left;
-        const height: i32 = bottom - top;
-        const destinationPitch: i32 = @divExact(rawPitch, 2);
 
-        var destination = @as([*]u16, @ptrFromInt(@intFromPtr(destinationBase) + @as(usize, @intCast(top)) * @as(usize, @intCast(rawPitch)) + @as(usize, @intCast(left)) * 2));
-
-        iy = 0;
-        while (iy < height) : (iy += 1) {
-            const rectPitch: i32 = iy * destinationPitch;
-
-            ix = 0;
-            while (ix < width) : (ix += 1) {
-                destination[@as(usize, @bitCast(rectPitch + ix))] = color;
+        pub fn line(start: [2]u8, end: [2]u8, color: Color) void {
+            // y always moves down
+            const p1, const p2 = if (start[1] < end[1])
+                .{ start, end }
+            else
+                .{ end, start };
+            const x1, const y1 = p1;
+            const x2, const y2 = p2;
+            //  Horizontal case
+            if (y1 == y2) {
+                lineHorizontal(@min(x1, x2), @max(x1, x2) + 1, y1, color);
+            //  Vertical case
+            } else if (x1 == x2) {
+                for (y1..y2 + 1) |y| setPixel(x1, @truncate(y), color);
+            } else {
+                var diff: i16 = 0;
+                var x, var y = p1;
+                const dx, const x_step: u8 = if (x1 < x2)
+                    .{ x2 - x1, 1 }
+                else
+                    .{ x1 - x2, @bitCast(@as(i8, -1)) };
+                const dy = y2 - y1;
+                while (true) {
+                    setPixel(x, y, color);
+                    if (x == x2 and y == y2)
+                        break;
+                    if (diff < 0) {
+                        x +%= x_step;
+                        diff += dy;
```

Comment:
This is now the only `@bitCast` in the `line` code, used only with explicitly wrapping math very locally.